ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt

```diff
@@ -31,13 +48,26 @@ class HttpRedirect {
             if (!origin.response.status.isRedirect()) return origin
 
             var call = origin
+            val originProtocol = origin.request.url.protocol
+            val originAuthority = origin.request.url.authority
             while (true) {
                 val location = call.response.headers[HttpHeaders.Location]
 
                 val requestBuilder = HttpRequestBuilder().apply {
                     takeFrom(origin.request)
                     url.parameters.clear()
 
+                    /**
+                     * Disallow redirect with a security downgrade.
+                     */
+                    if (originProtocol.isSecure() && !url.protocol.isSecure()) {
+                        return call
+                    }
+
+                    if (originAuthority != url.authority) {
```

Comment:
Should we only check for auth/host/port or we need to check path as well?

Reply:
Path is ok. We should only check for `authority` part of the url: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Generic_syntax